contracts/multisig/src/signing.rs

```diff
@@ -79,15 +77,32 @@ fn signers_weight(
         .sum()
 }
 
+fn update_session_state(
+    store: &mut dyn Storage,
+    session: &mut SigningSession,
+    key: &Key,
+    block_height: u64,
+) -> StdResult<()> {
+    // weight must be loaded outside `if`, otherwise it could cause gas cost inconsistency among signers
+    let weight = signers_weight(store, session.id.u64(), &key.snapshot)?;
+    if session.state == MultisigState::Pending && weight >= key.snapshot.quorum.into() {
+        session.state = MultisigState::Completed {
+            completed_at: block_height,
+        };
+    }
+
+    SIGNING_SESSIONS.save(store, session.id.u64(), session)
+}
+
 pub fn sign(
```

Comment:
consider if it makes sense to follow this pattern (sketch)
```
let mut session = load_session(store)
session.sign(signature)
save_session(session)
```
or even better, keep the session immutable if that makes sense:
```
let session = load_session(store).sign(signature)
save_session(session)
```

Reply:
Agree with Chris. The code atm is quite procedural.

Reply:
did some restructuring, please take a look again